[CSS Zoom] Apply zoom factor to InsetEdge

nullhook opened the pull request:
#### a754f38639293cb43493891f37d249b508362c7c
<pre>
[CSS Zoom] Apply zoom factor to InsetEdge
<a href="https://bugs.webkit.org/show_bug.cgi?id=300777">https://bugs.webkit.org/show_bug.cgi?id=300777</a>
<a href="https://rdar.apple.com/162663056">rdar://162663056</a>

Reviewed by Antti Koivisto.

Apply evaluation-time zoom for InsetEdge properties (top/left/right/bottom).

Additionally, we compensate for page zoom during CSS to Style conversion
on all viewport units.

* LayoutTests/TestExpectations:

* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/floats/crashtests/float-rewind-parallel-flow-3-crash.html:

We hit a debug assert in RenderLayerCompositor::fixedLayerIntersectsViewport()
where the combination of column-count, 3D transforms (especially 90/270 rotations),
and non-unit zoom values is causing non-determinism.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301363">https://bugs.webkit.org/show_bug.cgi?id=301363</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::outOfFlowNonReplacedVerticalGeometry const):
(WebCore::Layout::FormattingGeometry::outOfFlowNonReplacedHorizontalGeometry const):
(WebCore::Layout::FormattingGeometry::outOfFlowReplacedVerticalGeometry const):
(WebCore::Layout::FormattingGeometry::outOfFlowReplacedHorizontalGeometry const):
(WebCore::Layout::FormattingGeometry::inFlowPositionedPositionOffset const):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.h:
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::ContentChangeObserver::isVisuallyHidden):
* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::insetBeforeValue const):
(WebCore::PositionedLayoutConstraints::insetAfterValue const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::relativePositionOffset const):
(WebCore::RenderBoxModelObject::computeStickyPositionConstraints const):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::InsetEdgeSharedAdaptor::computedValue const):
* Source/WebCore/style/values/position/StyleInset.h:
* Source/WebCore/style/values/primitives/StyleLengthResolution.cpp:
(WebCore::Style::adjustValueForPageZoom):
(WebCore::Style::computeNonCalcLengthDouble):

Canonical link: <a href="https://commits.webkit.org/302102@main">https://commits.webkit.org/302102@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fd0b22d2278fca7290a90780ccf4cf4c6153f67

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127979 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38805 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135350 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79501 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5c3882d1-bf29-4789-b72d-f40955efc958) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129851 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/128 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97424 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65319 "") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9e5aaeeb-e9e2-4268-bf2f-ae832170eb52) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/81 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77991 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/95e51804-e709-4358-adab-bf459723bcb9) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78660 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108445 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33224 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137834 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/119 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105952 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/144 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110992 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105707 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/84 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52268 "Hash 2fd0b22d for PR 52371 does not build (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20005 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/164 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/61620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/92 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/168 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/135 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -7592,13 +7592,10 @@ imported/w3c/web-platform-tests/css/css-viewport/zoom/border.html [ ImageOnlyFai
 imported/w3c/web-platform-tests/css/css-viewport/zoom/bottom.html [ ImageOnlyFailure ]
```
I think bottom can be unskipped as well

anttijk commented on Source/WebCore/style/values/primitives/StyleLengthResolution.cpp:
```diff
@@ -41,10 +41,23 @@
+    if (!style || !evaluationTimeZoomEnabled(*style))
+        return dimension;
+
+    return dimension / style->pageZoomFactor();
```
Why does the pageZoomFactor need to be part of style? This appears to be the only client and it should be available here via `conversionData.renderView()->zoomFactor()`.

nullhook replied:
didn't see pageZoomFactor was exposed on render view. perhaps, we should rename it to more obvious `RenderView::pageZoomFactor` separately.

anttijk commented on Source/WebCore/style/values/primitives/StyleLengthResolution.cpp:
```diff
@@ -286,19 +299,19 @@ double computeNonCalcLengthDouble(double value, CSS::LengthUnit lengthUnit, cons
+        return value * adjustValueForPageZoom(lengthOfViewportPhysicalAxisForLogicalAxis(LogicalBoxAxis::Block, conversionData.defaultViewportFactor(), conversionData.style()), conversionData);
 
     case Vi:
         return value * lengthOfViewportPhysicalAxisForLogicalAxis(LogicalBoxAxis::Inline, conversionData.defaultViewportFactor(), conversionData.style());
```
Don't all viewport units need similar adjustment? Are you adding those separately?

nullhook replied:
added adjustment to all viewport units.

anttijk commented on LayoutTests/imported/w3c/web-platform-tests/css/CSS2/floats/crashtests/float-rewind-parallel-flow-3-crash.html:
```diff
@@ -1,3 +1,5 @@
+<!-- webkit-test-runner [ EvaluationTimeZoomEnabled=false ] -->
+<!-- FIXME: https://bugs.webkit.org/show_bug.cgi?id=301363 -->
 <!DOCTYPE html>
 <link rel="help" href="https://crbug.com/1520442">
 <style>
```
ok but it would nice to figure this one out (possibly by fixing the assert)

nullhook replied:
@mattwoodrow has a tangential fix that makes this crash go away. will resolve after this merges.

nullhook committed:
[CSS Zoom] Apply zoom factor to InsetEdge
https://bugs.webkit.org/show_bug.cgi?id=300777
rdar://162663056

Reviewed by Antti Koivisto.

Apply evaluation-time zoom for InsetEdge properties (top/left/right/bottom).

Additionally, we compensate for page zoom during CSS to Style conversion
on all viewport units.

* LayoutTests/TestExpectations:

* LayoutTests/imported/w3c/web-platform-tests/css/CSS2/floats/crashtests/float-rewind-parallel-flow-3-crash.html:

We hit a debug assert in RenderLayerCompositor::fixedLayerIntersectsViewport()
where the combination of column-count, 3D transforms (especially 90/270 rotations),
and non-unit zoom values is causing non-determinism.
https://bugs.webkit.org/show_bug.cgi?id=301363

* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/svg-computed-style.html:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::outOfFlowNonReplacedVerticalGeometry const):
(WebCore::Layout::FormattingGeometry::outOfFlowNonReplacedHorizontalGeometry const):
(WebCore::Layout::FormattingGeometry::outOfFlowReplacedVerticalGeometry const):
(WebCore::Layout::FormattingGeometry::outOfFlowReplacedHorizontalGeometry const):
(WebCore::Layout::FormattingGeometry::inFlowPositionedPositionOffset const):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.h:
* Source/WebCore/page/ios/ContentChangeObserver.cpp:
(WebCore::ContentChangeObserver::isVisuallyHidden):
* Source/WebCore/rendering/PositionedLayoutConstraints.h:
(WebCore::PositionedLayoutConstraints::insetBeforeValue const):
(WebCore::PositionedLayoutConstraints::insetAfterValue const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::relativePositionOffset const):
(WebCore::RenderBoxModelObject::computeStickyPositionConstraints const):
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::InsetEdgeSharedAdaptor::computedValue const):
* Source/WebCore/style/values/position/StyleInset.h:
* Source/WebCore/style/values/primitives/StyleLengthResolution.cpp:
(WebCore::Style::adjustValueForPageZoom):
(WebCore::Style::computeNonCalcLengthDouble):

Canonical link: https://commits.webkit.org/302102@main
